### PR TITLE
[RFC][OptionsResolver] POC for OptionsAwareTrait

### PR DESCRIPTION
**Pull request by ro0NL**
| Q | A |
| --- | --- |
| Branch? | "master" |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | ~ |
| Fixed tickets | ~ |
| License | MIT |
| Doc PR | ~ |

A quick proof of concept for a `OptionsAwareTrait`, basicly it a allows for specialized option classes. I came up with it after thinking about https://github.com/symfony/symfony/issues/18628#issuecomment-215334941

``` php
class MyOptions
{
    use OptionsAwareTrait {
        initOptions as public __construct; 
    }

    public function setFoo($value)
    {
        $this->setOption('foo', $value);
    }

    public function getFoo()
    {
        return $this->getOption('foo');
    }

    protected function configureOptions(OptionsResolver $resolver)
    {
        $resolver->setDefault('foo', 'default');
    }
}

$options = new MyOptions(array('foo' => 'bar'));
$options->getFoo(); // "bar"
```
- [x] POC
- [x] Discuss
- [ ] Tests
- [ ] Docs

### CODE REVIEW

**Comment by iltar** [comment]
As soon as I see magic methods becoming part of the public API, I turn around and run away screaming

**Comment by WouterJ** [comment]
@iltar as you can see, a magic one and a normal one is submitted by this PR :)

**Comment by iltar** [comment]
A :-1: would not be sufficient enough to say no to magic methods like this :cry: 

The normal implementation I can see being re-used. It might reduce a slight overhead of custom (repeated) code in implementations.

File: src/Symfony/Component/OptionsResolver/OptionsAwareTrait.php [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+/**
+ * Exposes option utilities for generic purposes.
+ *
+ * @author Roland Franssen <franssen.roland@gmail.com>
+ */
+trait OptionsAwareTrait
+{
+    private $strictOptions = true;
+    private $defaultOptionValue = null;
```
Comment (phansys) [inline src/Symfony/Component/OptionsResolver/OptionsAwareTrait.php]: `null` assignment isn't required as it is already the default value.

File: src/Symfony/Component/OptionsResolver/MagicOptionsAwareTrait.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+     * @throws \BadMethodCallException If the method is not a supported option method
+     */
+    public function __call($method, $arguments)
+    {
+        // @TODO allow for arbitrary options if $strictOptions is disabled?
+        if (preg_match('~^(is|has|get|set)(.+)$~i', $method, $matches)) {
+            $modifier = strtolower($matches[1]);
+            array_unshift($arguments, lcfirst($matches[2]));
+            try {
+                $cast = ($modifier === 'is' || $modifier === 'has');
```
Comment (phansys) [inline src/Symfony/Component/OptionsResolver/MagicOptionsAwareTrait.php]: Please, use Yoda conditions.

File: src/Symfony/Component/OptionsResolver/OptionsAwareTrait.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        if (null === $this->optionsResolver) {
+            $this->optionsResolver = new OptionsResolver();
+            $this->configureOptions($this->optionsResolver);
+        }
+
+        return $this->optionsResolver;
+    }
+
+    abstract protected function configureOptions(OptionsResolver $resolver);
```
Comment (phansys) [inline src/Symfony/Component/OptionsResolver/OptionsAwareTrait.php]: Please, place protected methods before private ones.

**Comment by ro0NL** [comment]
Reconsidered it a bit.. and i think we should remove the magic trait, $defaultOptionValue and $strictOptions to keep it really tight and clean. You can do that kinda stuff yourself anyway..

**Comment by xabbuh** [comment]
Not sure if I understand how you would use it in practice or what would be the advantage over how you do currently configure your options. Can you elaborate this a bit more?

**Comment by ro0NL** [comment]
@xabbuh it just a trait providing `setOption(s)` and `getOption(s)` powered by the `OptionsResolver`, ie. clear option definitions and validation.

The implementator has full control over exposing these methods yes/no. Private by default (see first example) but can also be implemented as public API, i.e.

```php
use OptionsAwareTrait {
    initOptions as public __construct;
    setOptions as public;
    setOption as public;
    getOptions as public;
    getOption as public;
}
```

`configureOptions` is the method to be implemented, always!

**Comment by xabbuh** [comment]
To be honest I am not sure if this is a common enough use case to be part of the core.

**Comment by ro0NL** [comment]
Understand, however i've seen tons of classes with `set/getOption(s)` method over the last decade. Besides that, symfony already provides a workflow to deal with "options", the `OptionsResolver` component and from that pov i think it's a nice addition. We enforce users to make use of the options resolver this way whenever they need this kind of methods for whatever options, which is the true added value imo.

**Comment by ro0NL** [comment]
Closing as i get no clear yes/no to finish this. Assuming it's not needed/wanted.

**Pull request closed without merging**